File: frontend/src/lib/algorithms/bfs.ts
import { TransitGraph } from "@/lib/transit-graph";
import { LINE_LOOKUP } from "@/lib/transit-graph";
import { type RouteSearchResult, type PathEntry, makeSegment } from "./types";

export function bfsSearch(
  graph: TransitGraph,
  originId: string,
  destId: string,
): RouteSearchResult | null {
  if (graph.isNodeBlocked(originId) || graph.isNodeBlocked(destId)) return null;

  type QueueEntry = {
    station: string;
    line: string;
    transfers: number;
    path: PathEntry[];
    timeAcc: number;
    distAcc: number;
  };

  const queue: QueueEntry[] = [];
  const visited = new Set<string>();

  for (const edge of graph.getTraversableNeighbors(originId)) {
    const stateKey = `${originId}|${edge.line_id}`;
    if (!visited.has(stateKey)) {
      visited.add(stateKey);
      const effTime = graph.getEffectiveTravelTime(originId, edge);
      const destKey = `${edge.to_node_id}|${edge.line_id}`;
      visited.add(destKey);
      queue.push({
        station: edge.to_node_id,
        line: edge.line_id,
        transfers: 0,
        path: [
          [originId, "START"],
          [edge.to_node_id, edge.line_id],
        ],
        timeAcc: effTime,
        distAcc: edge.distance_km,
      });
    }
  }

  let bestResult: RouteSearchResult | null = null;

  let head = 0;
  while (head < queue.length) {
    const { station, line, transfers, path, timeAcc, distAcc } = queue[head++];

    if (station === destId) {
      const candidate = buildResult(graph, path, timeAcc, distAcc, transfers);
      if (bestResult === null || transfers < bestResult.total_transfers) {
        bestResult = candidate;
      } else if (
        transfers === bestResult.total_transfers &&
        timeAcc < bestResult.total_time_min
      ) {
        bestResult = candidate;
      }
      continue;
    }

    for (const edge of graph.getTraversableNeighbors(station)) {
      const newLine = edge.line_id;
      const isTransfer = newLine !== line;
      const newTransfers = transfers + (isTransfer ? 1 : 0);

      if (bestResult && newTransfers > bestResult.total_transfers) continue;

      const stateKey = `${edge.to_node_id}|${newLine}`;
      if (visited.has(stateKey)) continue;
      visited.add(stateKey);

      const effTime = graph.getEffectiveTravelTime(station, edge);
      queue.push({
        station: edge.to_node_id,
        line: newLine,
        transfers: newTransfers,
        path: [...path, [edge.to_node_id, newLine]],
        timeAcc: timeAcc + effTime,
        distAcc: distAcc + edge.distance_km,
      });
    }
  }

  return bestResult;
}

function buildResult(
  graph: TransitGraph,
  path: PathEntry[],
  totalTime: number,
  totalDist: number,
  transfers: number,
): RouteSearchResult {
  const segments = [];
  let segStart = 0;

  for (let i = 1; i < path.length; i++) {
    if (path[i][1] !== path[segStart][1] && path[segStart][1] !== "START") {
      segments.push(makeSegmentBfs(graph, path, segStart, i - 1));
      segStart = i - 1;
    } else if (path[segStart][1] === "START" && i > 0) {
      const segStartLine = path.length > 1 ? path[1][1] : "";
      if (i > 1 && path[i][1] !== segStartLine) {
        segments.push(makeSegmentBfs(graph, path, 0, i - 1));
        segStart = i - 1;
      }
    }
  }

  segments.push(makeSegmentBfs(graph, path, segStart, path.length - 1));

  const totalCo2 = segments.reduce((s, seg) => s + seg.co2_grams, 0);

  return {
    segments,
    total_time_min: Math.round(totalTime * 10) / 10,
    total_distance_km: Math.round(totalDist * 100) / 100,
    total_transfers: transfers,
    total_co2_grams: Math.round(totalCo2 * 10) / 10,
    visited_ids: path.map((p) => p[0]),
  };
}

function makeSegmentBfs(
  graph: TransitGraph,
  path: PathEntry[],
  start: number,
  end: number,
) {
  let lineId = path[end][1];
  if (lineId === "START" && end > start) {
    lineId = path[start + 1][1];
  }
  return makeSegment(graph, path, start, end, lineId);
}

File: backend/app/algorithms/bfs.py
"""
BFS — Minimum Transfers Route (Disruption-Aware).

Models the graph at the *line-segment* level: each state is (station, current_line).
BFS on transfer count finds the path that minimizes line changes.
Within equal-transfer paths, picks the shorter one by total travel time.

Blocked nodes/edges are excluded via ``get_traversable_neighbors``.
Travel times include incident penalties via ``get_effective_travel_time``.
Each result segment carries a ``co2_grams`` field.
"""

from __future__ import annotations

from collections import deque
from typing import Optional, TYPE_CHECKING

from app.data.chennai_network import LINES, LINE_LOOKUP

if TYPE_CHECKING:
    from app.data.chennai_network import TransitGraph


def bfs_search(
    graph: TransitGraph, origin_id: str, dest_id: str
) -> Optional[dict]:
    if graph.is_node_blocked(origin_id) or graph.is_node_blocked(dest_id):
        return None

    queue: deque[
        tuple[str, str, int, list[tuple[str, str]], float, float]
    ] = deque()
    visited: set[tuple[str, str]] = set()

    for edge in graph.get_traversable_neighbors(origin_id):
        state = (origin_id, edge.line_id)
        if state not in visited:
            visited.add(state)
            eff_time = graph.get_effective_travel_time(origin_id, edge)
            queue.append((
                edge.to_node_id,
                edge.line_id,
                0,
                [(origin_id, "START"), (edge.to_node_id, edge.line_id)],
                eff_time,
                edge.distance_km,
            ))
            visited.add((edge.to_node_id, edge.line_id))

    best_result: Optional[dict] = None

    while queue:
        station, line, transfers, path, time_acc, dist_acc = queue.popleft()

        if station == dest_id:
            candidate = _build_result(
                graph, path, time_acc, dist_acc, transfers
            )
            if best_result is None or transfers < best_result["total_transfers"]:
                best_result = candidate
            elif (
                transfers == best_result["total_transfers"]
                and time_acc < best_result["total_time_min"]
            ):
                best_result = candidate
            continue

        for edge in graph.get_traversable_neighbors(station):
            new_line = edge.line_id
            is_transfer = new_line != line
            new_transfers = transfers + (1 if is_transfer else 0)

            if best_result and new_transfers > best_result["total_transfers"]:
                continue

            state = (edge.to_node_id, new_line)
            if state in visited:
                continue
            visited.add(state)

            eff_time = graph.get_effective_travel_time(station, edge)
            queue.append((
                edge.to_node_id,
                new_line,
                new_transfers,
                path + [(edge.to_node_id, new_line)],
                time_acc + eff_time,
                dist_acc + edge.distance_km,
            ))

    return best_result


def _build_result(
    graph: TransitGraph,
    path: list[tuple[str, str]],
    total_time: float,
    total_dist: float,
    transfers: int,
) -> dict:
    segments: list[dict] = []
    seg_start = 0

    for i in range(1, len(path)):
        if path[i][1] != path[seg_start][1] and path[seg_start][1] != "START":
            segments.append(_make_segment(graph, path, seg_start, i - 1))
            seg_start = i - 1
        elif path[seg_start][1] == "START" and i > 0:
            seg_start_line = path[1][1] if len(path) > 1 else ""
            if i > 1 and path[i][1] != seg_start_line:
                segments.append(_make_segment(graph, path, 0, i - 1))
                seg_start = i - 1

    segments.append(_make_segment(graph, path, seg_start, len(path) - 1))

    total_co2 = sum(s["co2_grams"] for s in segments)

    return {
        "segments": segments,
        "total_time_min": round(total_time, 1),
        "total_distance_km": round(total_dist, 2),
        "total_transfers": transfers,
        "total_co2_grams": round(total_co2, 1),
        "visited_ids": [p[0] for p in path],
    }


def _make_segment(
    graph: TransitGraph,
    path: list[tuple[str, str]],
    start: int,
    end: int,
) -> dict:
    from_s = graph.get_station(path[start][0])
    to_s = graph.get_station(path[end][0])

    line_id = path[end][1]
    if line_id == "START" and end > start:
        line_id = path[start + 1][1]

    line_info = LINE_LOOKUP.get(line_id)
    mode = line_info.mode if line_info else "metro"

    total_time = 0.0
    total_dist = 0.0
    total_co2 = 0.0
    intermediates: list = []

    for i in range(start, end):
        sid = path[i][0]
        nid = path[i + 1][0]
        found = False
        for edge in graph.get_neighbors(sid):
            if edge.to_node_id == nid and edge.line_id == line_id:
                total_time += graph.get_effective_travel_time(sid, edge)
                total_dist += edge.distance_km
                total_co2 += edge.emission_gco2_per_km * edge.distance_km
                found = True
                break
        if not found:
            for edge in graph.get_neighbors(sid):
                if edge.to_node_id == nid:
                    total_time += graph.get_effective_travel_time(sid, edge)
                    total_dist += edge.distance_km
                    total_co2 += edge.emission_gco2_per_km * edge.distance_km
                    break

        if start < i < end:
            intermediates.append(graph.get_station(path[i][0]))

    return {
        "from_station": from_s,
        "to_station": to_s,
        "line": line_id,
        "mode": mode,
        "travel_time_min": round(total_time, 1),
        "distance_km": round(total_dist, 2),
        "co2_grams": round(total_co2, 1),
        "intermediate_stations": intermediates,
    }
